API Add a CREATE_MEMORY_HYDRATED option to DataObject constructor

maxime-rainville opened the pull request:
This adds a fourth option to the DataObject constructor to hydrate from memory.

Fixes https://github.com/silverstripe/silverstripe-framework/issues/9750

madmatt commented on src/ORM/DataObject.php:
```diff
@@ -203,6 +203,12 @@ class DataObject extends ViewableData implements DataObjectInterface, i18nEntity
 
+    /**
+     * Value for 2nd argument to constructor, indicating that a record is being hydrated from memory. This can be used
+     * to initialised a record that doesn't yet have an ID. Neither setters and nor default population will be called.
```
Sentence could be structured a bit more clearly - e.g. `Setter methods are not called, and population via private static $defaults will not occur`.

madmatt commented on src/ORM/DataObject.php:
```diff
@@ -339,7 +345,9 @@ public static function getSchema()
      * @param array $record Initial record content, or rehydrated record content, depending on $creationType
-     * @param int|boolean $creationType Set to DataObject::CREATE_OBJECT, DataObject::CREATE_HYDRATED, or DataObject::CREATE_SINGLETON. Used by SilverStripe internals as best left as the default by regular users.
+     * @param int|boolean $creationType Set to DataObject::CREATE_OBJECT, DataObject::CREATE_HYDRATED,
+     *   DataObject::CREATE_MEMORY_HYDRATED or DataObject::CREATE_SINGLETON. Used by Silverstripe internals as best
```
The first occurrence of `as` should be `and` here.

maxime-rainville commented on tests/php/ORM/DataObjectTest.php:
```diff
@@ -1669,6 +1669,31 @@ public function testNewClassInstance()
+
+        $this->assertInstanceOf(DataObjectTest\SubTeam::class, $changedDO);
+        $this->assertEquals($changedDO->ClassName, DataObjectTest\SubTeam::class);
+        $this->assertNotEmpty($changedDO->ID, );
```
```suggestion
        $this->assertNotEmpty(
            $changedDO->ID,
            'An ID has been generated for the new class instance on save'
        );
```

maxime-rainville commented:
This didn't make the cut off for the 4.7 release. We're thinking we might be able to call it a bug fix since it's technically not a an API change. This would allow us to ship it in 4.7.1. We'll sort it out, once 4.7.0 stable is out the door.

emteknetnz reviewed (changes requested):
There's a lot of duplicated code between the regular hydration and the memory hydration. If we do a bugfix sometime in the future on one of them we might forget to fix the other one. Should create a private method with shared code.

sminnee commented:
Or use have both options branch to the same case block, and do a further conditional/ternary inside that for the 1 difference.

maxime-rainville commented:
Addressed @emteknetnz's feedback. I moved the hydration logic into a private `hydrate` method with a flag that account for the difference between CREATE_HYDRATED and CREATE_MEMORY_HYDRATED.

KhushbuFuletra commented:
Hi, I am waiting for this one. 
when do you think you will be able to tag this PR?

KhushbuFuletra commented:
> It's going to be in the 4.8.0 release which is still a couple months away.

Thanks, @maxime-rainville  for the heads-up.
